src/compiler/program.ts

```diff
@@ -151,18 +151,25 @@ module ts {
         let files: SourceFile[] = [];
         let filesByName: Map<SourceFile> = {};
         let diagnostics = createDiagnosticCollection();
-        let seenNoDefaultLib = options.noLib;
         let commonSourceDirectory: string;
         let diagnosticsProducingTypeChecker: TypeChecker;
         let noDiagnosticsTypeChecker: TypeChecker;
 
+        // shouldExcludeDefaultLib is true if:
+        //  - The '--noLib' flag is used.
+        //  - A 'no-default-lib' reference comment is encountered in
+        //      processing the root files.
+        let shouldExcludeDefaultLib = options.noLib;

```

Comment:
This seems unecessary.  why not have a "let sawNoDefaultLibTag" which you set while processing.  Then , have your check be:

```if (!sawNoDefaultLib && !options.noLib) { ... }```

Merging the two disparate systems into one variable, which you mutate on the fly seems much harder to understand.